<a href="https://colab.research.google.com/github/MEYLIN84/DATA-UPC/blob/main/Applied_Capstone_Week_3_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# **Agrupación y Segmentación de Vecindarios en la Ciudad de Toronto, Canadá**

In [1]:
import pandas as pd

PREGUNTA 1

Se utiliza una versión anterior de la página web de wikipedia, debido a que al actual tiene un cambio al presentar la tabla de códigos postales.

In [2]:
import requests

In [7]:
url = "https://en.m.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&oldid=945633050"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

La respuesta 200 significa que se ha conectado a la página correctamente

In [8]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postcode           Borough          Neighbourhood
 0        M1A      Not assigned           Not assigned
 1        M2A      Not assigned           Not assigned
 2        M3A        North York              Parkwoods
 3        M4A        North York       Victoria Village
 4        M5A  Downtown Toronto           Harbourfront
 ..       ...               ...                    ...
 282      M8Z         Etobicoke              Mimico NW
 283      M8Z         Etobicoke     The Queensway West
 284      M8Z         Etobicoke  Royal York South West
 285      M8Z         Etobicoke         South of Bloor
 286      M9Z      Not assigned           Not assigned
 
 [287 rows x 3 columns]]

In [9]:
len(wiki_data), type(wiki_data)

(1, list)

Necesito solo la primera tabla por lo que se eliminan las demás.

In [10]:
wiki_data = wiki_data[0]
wiki_data

Postcode           Borough          Neighbourhood
0        M1A      Not assigned           Not assigned
1        M2A      Not assigned           Not assigned
2        M3A        North York              Parkwoods
3        M4A        North York       Victoria Village
4        M5A  Downtown Toronto           Harbourfront
..       ...               ...                    ...
282      M8Z         Etobicoke              Mimico NW
283      M8Z         Etobicoke     The Queensway West
284      M8Z         Etobicoke  Royal York South West
285      M8Z         Etobicoke         South of Bloor
286      M9Z      Not assigned           Not assigned

[287 rows x 3 columns]

Se elimina los registros que tiene NOT ASSIGNED en la columna Borough 

In [11]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postcode           Borough             Neighbourhood
2        M3A        North York                 Parkwoods
3        M4A        North York          Victoria Village
4        M5A  Downtown Toronto              Harbourfront
5        M6A        North York          Lawrence Heights
6        M6A        North York            Lawrence Manor
..       ...               ...                       ...
281      M8Z         Etobicoke  Kingsway Park South West
282      M8Z         Etobicoke                 Mimico NW
283      M8Z         Etobicoke        The Queensway West
284      M8Z         Etobicoke     Royal York South West
285      M8Z         Etobicoke            South of Bloor

[210 rows x 3 columns]

Se agrupan los registros según el código postal (postcode)

In [13]:
df = df.groupby(['Postcode']).head()
df

Postcode           Borough             Neighbourhood
2        M3A        North York                 Parkwoods
3        M4A        North York          Victoria Village
4        M5A  Downtown Toronto              Harbourfront
5        M6A        North York          Lawrence Heights
6        M6A        North York            Lawrence Manor
..       ...               ...                       ...
281      M8Z         Etobicoke  Kingsway Park South West
282      M8Z         Etobicoke                 Mimico NW
283      M8Z         Etobicoke        The Queensway West
284      M8Z         Etobicoke     Royal York South West
285      M8Z         Etobicoke            South of Bloor

[202 rows x 3 columns]

Se verifica el número de registros con el valor Not assigned en la columna Neighbourhood

In [14]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [15]:
df = df.reset_index()
df

index Postcode           Borough             Neighbourhood
0        2      M3A        North York                 Parkwoods
1        3      M4A        North York          Victoria Village
2        4      M5A  Downtown Toronto              Harbourfront
3        5      M6A        North York          Lawrence Heights
4        6      M6A        North York            Lawrence Manor
..     ...      ...               ...                       ...
197    281      M8Z         Etobicoke  Kingsway Park South West
198    282      M8Z         Etobicoke                 Mimico NW
199    283      M8Z         Etobicoke        The Queensway West
200    284      M8Z         Etobicoke     Royal York South West
201    285      M8Z         Etobicoke            South of Bloor

[202 rows x 4 columns]

In [16]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postcode           Borough             Neighbourhood
0        M3A        North York                 Parkwoods
1        M4A        North York          Victoria Village
2        M5A  Downtown Toronto              Harbourfront
3        M6A        North York          Lawrence Heights
4        M6A        North York            Lawrence Manor
..       ...               ...                       ...
197      M8Z         Etobicoke  Kingsway Park South West
198      M8Z         Etobicoke                 Mimico NW
199      M8Z         Etobicoke        The Queensway West
200      M8Z         Etobicoke     Royal York South West
201      M8Z         Etobicoke            South of Bloor

[202 rows x 3 columns]

In [17]:
df.shape

(202, 3)

Respuesta: El dataframe tiene 202 filas y 3 columnas. 

PREGUNTA 2

Se instala el geocoder

In [18]:
pip install geocoder

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 98 kB 7.0 MB/s 


In [19]:
import geocoder # import geocoder

Se importa el archivo csv desde la URL.

In [20]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [22]:
print("El tamaño de la data de wiki es: ", df.shape)
print("El tamaño de la data de csv es: ", data.shape)

El tamaño de la data de wiki es:  (202, 3)
El tamaño de la data de csv es:  (103, 3)


Se verifica los tipos de datos de los dos dataframes.



In [23]:
df.dtypes

Postcode         object
Borough          object
Neighbourhood    object
dtype: object

In [24]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [29]:
combined_data = df.join(data.set_index('Postal Code'), on='Postcode', how='inner')
combined_data

Postcode           Borough             Neighbourhood   Latitude  Longitude
0        M3A        North York                 Parkwoods  43.753259 -79.329656
1        M4A        North York          Victoria Village  43.725882 -79.315572
2        M5A  Downtown Toronto              Harbourfront  43.654260 -79.360636
3        M6A        North York          Lawrence Heights  43.718518 -79.464763
4        M6A        North York            Lawrence Manor  43.718518 -79.464763
..       ...               ...                       ...        ...        ...
197      M8Z         Etobicoke  Kingsway Park South West  43.628841 -79.520999
198      M8Z         Etobicoke                 Mimico NW  43.628841 -79.520999
199      M8Z         Etobicoke        The Queensway West  43.628841 -79.520999
200      M8Z         Etobicoke     Royal York South West  43.628841 -79.520999
201      M8Z         Etobicoke            South of Bloor  43.628841 -79.520999

[202 rows x 5 columns]

In [30]:
combined_data.shape

(202, 5)

Respuesta: El dataframe tiene 202 filas y 5 columnas, se asignó correctamente los datos provenientes del csv.

PREGUNTA 3

In [31]:
from geopy.geocoders import Nominatim 

In [35]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('Las coordenadas de Toronto son  {}, {}.'.format(latitude, longitude))

Las coordenadas de Toronto son  43.6534817, -79.3839347.


MAPA DE TORONTO

In [33]:
import folium

In [36]:
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto